Source/WebCore/Modules/url-pattern/URLPattern.cpp

```diff
@@ -26,29 +26,239 @@
+        }
+    }
+
+    if (!init.protocol.isNull()) {
+        ExceptionOr<String> protocolResult = canonicalizeProtocol(WTFMove(init.protocol), type);
+
+        if (protocolResult.hasException())
+            return protocolResult.releaseException();
+
+        result.protocol = protocolResult.releaseReturnValue();
+    }
+
+    if (!init.username.isNull())
+        result.username = canonicalizeUsername(WTFMove(init.username), type);
+
+    if (!init.password.isNull())
+        result.password = canonicalizePassword(WTFMove(init.password), type);
+
+    if (!init.hostname.isNull()) {
+        ExceptionOr<String> hostResult = canonicalizeHost(WTFMove(init.protocol), type);
+
+        if (hostResult.hasException())
+            return hostResult.releaseException();
+
+        result.hostname = hostResult.releaseReturnValue();
+    }
+
+    if (!init.port.isNull()) {
+        ExceptionOr<String> portResult = canonicalizePort(WTFMove(init.port), WTFMove(init.protocol), type);
+
+        if (portResult.hasException())
+            return portResult.releaseException();
+
+        result.port = portResult.releaseReturnValue();
+    }
+
+    if (!init.pathname.isNull()) {
+        result.pathname = init.pathname;
+
+        if (!baseURL.isNull() && baseURL.hasOpaquePath() && !isAbsolutePathname(result.pathname, type)) {
+            String baseURLPath = processBaseURLString(baseURL.path(), type);
+            size_t slashIndex = baseURLPath.reverseFind("/"_s);
+
+            if (slashIndex != WTF::notFound)
+                result.pathname = makeString(baseURLPath.substring(0, slashIndex + 1), result.pathname);
+
+            ExceptionOr<String> pathResult = canonicalizePath(WTFMove(result.pathname), baseURL.protocol(), type);
+
+            if (pathResult.hasException())
+                return pathResult.releaseException();
+
+            result.pathname = pathResult.releaseReturnValue();
+        }
+    }
+
+    if (!init.search.isNull()) {
+        ExceptionOr<String> queryResult = canonicalizeSearch(WTFMove(init.search), type);
+
+        if (queryResult.hasException())
+            return queryResult.releaseException();
+
+        result.search = queryResult.releaseReturnValue();
+    }
+
+    if (!init.hash.isNull()) {
+        ExceptionOr<String> fragmentResult = canonicalizeHash(WTFMove(init.hash), type);
+
+        if (fragmentResult.hasException())
+            return fragmentResult.releaseException();
+
+        result.hash = fragmentResult.releaseReturnValue();
+    }
+
+    return result;
+}
+
+ExceptionOr<Ref<URLPattern>> URLPattern::create(URLPatternInput&&, String&& baseURL, URLPatternOptions&&)
 {
     UNUSED_PARAM(baseURL);
 
     return adoptRef(*new URLPattern());
 }
 
-Ref<URLPattern> URLPattern::create(std::optional<URLPatternInput>&&, URLPatternOptions&&)
+ExceptionOr<Ref<URLPattern>> URLPattern::create(std::optional<URLPatternInput>&& argInput, URLPatternOptions&&)
```

Comment:
`input`. `argInput` is not a good name as we don't like abbreviations in WebKit.